[NVIDIA XLA] NFC Unify GPU backend configs

Tixxx opened the pull request:
This PR unifies all GPU backend configs into a single one, this is to make it easy to add generic attributes in the future.

frgossen reviewed (approved):
Thanks!

cezheng commented:
This is probably against the principle to put backend specific fields to BackendConfig, because we don't want fields only used by one backend to bloat sizeof(HloInstruction), which is something a few folks have been spending some effort to reduce. Would you please clean this up?

Tixxx commented:
> This is probably against the principle to put backend specific fields to BackendConfig, because we don't want fields only used by one backend to bloat sizeof(HloInstruction), which is something a few folks have been spending some effort to reduce. Would you please clean this up?

Thanks for raising this issue. The idea of introducing this field is to support running instructions in multiple queues in general, backends can interpret it in any way they desire similar to the precision_config field. If adding it to hloInstruction proto is not the best route, could you suggest an alternative?

cezheng commented:
The idiomatic location is to put it in GPU's [BackendConfig](https://github.com/openxla/xla/blob/72d001249852ebe1d4e29afc4bc5b583fd07a2d0/xla/service/gpu/backend_configs.proto), one example commit to remove similar fields and move to CPU's BackendConfig is https://github.com/openxla/xla/commit/99af7933ab1197c48b62497d0f0cc88e4a750be4

cheshire commented:
Agreed with @cezheng , @Tixxx would you follow up? Otherwise we might need a rollback.

Tixxx commented:
> Agreed with @cezheng , @Tixxx would you follow up? Otherwise we might need a rollback.

Yea, I'm already working on changes based on @cezheng 's suggestion. I will update it early next week.

Tixxx commented:
@cheshire @cezheng @frgossen  I updated this PR to move the newly added attributes under xla/gpu configs. Please take a look.

cheshire commented on xla/service/gpu/backend_configs.proto:
```diff
@@ -197,3 +197,18 @@ message CudnnfMHABackendConfig {
 }
+
+// Generic backend config for XLA:GPU
+message GpuBackendConfig {
```
Would it clash with other backend configs we attach to instructions? E.g. using it to store tiling?

cezheng replied:
+1. It might make sense to make GPU backend config the single entry point that potentially contains other existing backend config types as members similar to other backends, but that is a larger change.

Tixxx replied:
looks like any new config will overwrite the existing config so there's a good chance that the GpuBackendConfig will be lost or vice versa. I think @cezheng 's suggestion is a good long-term solution but it'd take significantly more effort. For this PR, can I put these 2 fields in metadata temporarily? I will add a TODO and work on a followup pr to refactor various gpu backend configs and have a single entry.

Tixxx replied:
Any comments on the metadata route? @cheshire @cezheng

cheshire replied:
Metadata isn't great either, this is meant for, well, metadata. Maybe create new backend config to unify previous one + whatever data you need?

Tixxx replied:
fyi, i have been working on unifying all configs into a single GpuBackendConfig. still fixing some local test failures, but should have a new update soon.

Tixxx replied:
refactored all gpu backend configs into a single GpuBackendConfig

cheshire reviewed (changes requested):
Makes sense overall, but would need to see how it interacts with existing backend_config usage.

cezheng commented on xla/service/gpu/backend_configs.proto:
```diff
@@ -197,3 +201,32 @@ message CudnnfMHABackendConfig {
+  // multiple operation queues.
+  repeated int64 wait_on_operation_queues = 2;
+
+  CudnnConvBackendConfig cudnn_conv_backend_config = 3;
+
+  GemmBackendConfig gemm_backend_config = 4;
+
+  BitcastBackendConfig bitcast_backend_config = 5;
+
+  CollectiveBackendConfig collective_backend_config = 6;
+
+  FusionBackendConfig fusion_backend_config = 7;
+
+  CudnnNormBackendConfig cudnn_norm_backend_config = 8;
+
+  CudnnfMHABackendConfig cudnn_fmha_backend_config = 9;
```
nit: would it make sense to put all these into a `oneof`? Assuming these are already mutually exclusive in the existing code, but I'm not sure if they will ever coexist, @cheshire would be the better person to make the call.

Tixxx replied:
i think it makes sense to make it a oneof, they were meant to be individual configs for a given library or a codegen. i changed it to be oneof.

cezheng commented on xla/service/gpu/conv_algorithm_picker.cc:
```diff
@@ -1111,12 +1111,14 @@ StatusOr<bool> GpuConvAlgorithmPicker::RunOnInstruction(HloInstruction* instr) {
-  TF_ASSIGN_OR_RETURN(CudnnConvBackendConfig backend_config,
-                      instr->backend_config<CudnnConvBackendConfig>());
+  TF_ASSIGN_OR_RETURN(GpuBackendConfig gpu_backend_config,
+                      instr->backend_config<GpuBackendConfig>());
+  CudnnConvBackendConfig backend_config =
+      gpu_backend_config.cudnn_conv_backend_config();
   *backend_config.mutable_algorithm() = best_algo.algorithm();
   backend_config.mutable_algorithm()->mutable_workspace_size()->set_value(
       best_algo.scratch_bytes());
-
+  *gpu_backend_config.mutable_cudnn_conv_backend_config() = backend_config;
```
instead of making a deep copy, modifying then assigning back, you could just do

```
CudnnConvBackendConfig& backend_config = *gpu_backend_config.cudnn_conv_backend_config();
...
```
and delete the assignment by the end.

cezheng commented on xla/service/gpu/cudnn_fused_conv_rewriter.cc:
```diff
@@ -331,8 +333,8 @@ StatusOr<bool> FuseConvAlpha(HloComputation* comp) {
     config.set_conv_result_scale(alpha_f64.GetFirstElement<double>());
-
-    TF_RETURN_IF_ERROR(conv->set_backend_config(config));
+    *gpu_config.mutable_cudnn_conv_backend_config() = config;
```
ditto and same applies to all other applicable locations (not commenting on every single one since there are a lot)

cezheng commented on xla/service/gpu/cudnn_fused_conv_rewriter.cc:
```diff
@@ -1406,21 +1423,21 @@ void VlogStats(HloModule* module) {
       }
-
-      if (config->conv_result_scale() != 1) {
+      CudnnConvBackendConfig config = gpu_config->cudnn_conv_backend_config();
```
use const reference to avoid a deep copy

cezheng commented on xla/service/gpu/cudnn_fused_mha_rewriter.cc:
```diff
@@ -1331,7 +1331,9 @@ StatusOr<HloInstruction*> FuseFwdMultiHeadedAttentionBlock(
           call_shape, operands, absl::string_view(custom_call_name)));
-  TF_RETURN_IF_ERROR(fmha_call->set_backend_config(fmha_config));
+  GpuBackendConfig gpu_config;
+  *gpu_config.mutable_cudnn_fmha_backend_config() = fmha_config;
```
ditto for this and the other occurrence as I commented in xla/service/gpu/conv_algorithm_picker.cc

cezheng commented on xla/service/gpu/cudnn_fused_mha_transpose_fusion.cc:
```diff
@@ -271,8 +272,8 @@ StatusOr<bool> FuseArgPrologueTransposeWithcuDNNFMHA(
 
-  fmha->clear_backend_config();
-  TF_RETURN_IF_ERROR(fmha->set_backend_config(new_fmha_config));
+  *gpu_config.mutable_cudnn_fmha_backend_config() = new_fmha_config;
```
ditto for this and other occurrences as I commented in xla/service/gpu/conv_algorithm_picker.cc

cezheng commented on xla/service/gpu/cudnn_norm_rewriter.cc:
```diff
@@ -628,11 +630,14 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
       backend_config.mutable_algorithm()->mutable_workspace_size()->set_value(
           workspace_size);
-      TF_RETURN_IF_ERROR(custom_call->set_backend_config(backend_config));
+      *gpu_config.mutable_cudnn_norm_backend_config() = backend_config;
```
ditto

cezheng commented on xla/service/gpu/cudnn_vectorize_convolutions.cc:
```diff
@@ -289,10 +289,12 @@ Status ReorderInt8NchwVect(HloCustomCallInstruction* conv, XlaOp* operands) {
+  CudnnConvBackendConfig config = gpu_config.cudnn_conv_backend_config();
   config.set_reordered_int8_nchw_vect(true);
-  TF_RETURN_IF_ERROR(conv->set_backend_config(config));
+  *gpu_config.mutable_cudnn_conv_backend_config() = config;
```
ditto

cezheng commented on xla/service/gpu/custom_fusion_rewriter.cc:
```diff
@@ -166,7 +166,9 @@ static StatusOr<HloInstruction*> CreateFusionInstruction(
   backend_config.set_kind("__custom_fusion");
   *backend_config.mutable_custom_fusion_config() = match.config();
-  TF_RETURN_IF_ERROR(fusion->set_backend_config(std::move(backend_config)));
+  GpuBackendConfig gpu_config;
```
similar but you could just move the `GpuBackendConfig gpu_config;` declaration above, and do

```
FusionBackendConfig& backend_config = *gpu_config.mutable_fusion_backend_config();
```

then drop the assignment by the end. There are other occurrences in other files as well.

cezheng commented on xla/service/gpu/elemental_ir_emitter.cc:
```diff
@@ -158,8 +158,11 @@ llvm_ir::IrArray::Index GpuElementalIrEmitter::GetSourceIndexOfBitcast(
+  auto gpu_config = hlo->backend_config<GpuBackendConfig>();
+  CHECK(gpu_config.ok());
+
+  BitcastBackendConfig bitcast_config =
```
nit: const reference

cezheng commented on xla/service/gpu/gemm_algorithm_picker.cc:
```diff
@@ -420,7 +423,8 @@ StatusOr<bool> RunOnInstruction(HloInstruction* gemm,
     }
   }
-  TF_RETURN_IF_ERROR(gemm->set_backend_config(updated_config));
+  *gpu_config.mutable_gemm_backend_config() = updated_config;
```
ditto

cezheng commented on xla/service/gpu/gemm_rewriter.cc:
```diff
@@ -1045,9 +1053,8 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
-
-    TF_RETURN_IF_ERROR(
-        new_custom_call->set_backend_config(gemm_backend_config));
+    *gpu_backend_config.mutable_gemm_backend_config() = gemm_backend_config;
```
ditto

cezheng commented on xla/service/gpu/gemm_rewriter_triton.cc:
```diff
@@ -631,10 +631,12 @@ class GemmRewriterTritonVisitor : public DfsHloRewriteVisitor {
+    FusionBackendConfig backend_config = gpu_config.fusion_backend_config();
     backend_config.set_kind(std::string(kTritonGemmFusionKind));
-    TF_RETURN_IF_ERROR(dot_fusion->set_backend_config(backend_config));
+    *gpu_config.mutable_fusion_backend_config() = backend_config;
```
ditto

cezheng reviewed (changes requested):
Thank you for the cleanup! Looks good overall, but there are some nits that are repeated many times so I didn't end up commenting on every single instance, would you please make a global cleanup for that, thanks again.

Tixxx commented:
> Thank you for the cleanup! Looks good overall, but there are some nits that are repeated many times so I didn't end up commenting on every single instance, would you please make a global cleanup for that, thanks again.

addressed the nits, thanks for the review. Please take another look to see if there are issues.

cezheng reviewed (approved):
Thank you!

cheshire reviewed (changes requested):
Could we separate an NFC PR unifying backend configs?

Tixxx commented:
> Could we separate an NFC PR unifying backend configs?

I have removed changes related to the new attribute. I will open a separate PR to move the attributes from hloInstruction to GPUBackendConfig. Can you review this when you a chance?

cheshire commented on xla/service/collective_permute_decomposer.cc:
```diff
@@ -77,8 +77,9 @@ bool ShouldDecompose(const HloCollectivePermuteInstruction& collective_permute,
-      collective_permute.backend_config<xla::gpu::CollectiveBackendConfig>()
-          .value();
+      collective_permute.backend_config<xla::gpu::GpuBackendConfig>()
+          .value()
```
Super-nit: could also do `->collective_backend_config()` ?

cheshire commented on xla/service/gpu/buffer_sharing.cc:
```diff
@@ -210,8 +210,9 @@ std::optional<bool> CanShareBufferHint(const HloInstruction* user,
-            std::move(user->backend_config<GemmBackendConfig>()).value();
+        GpuBackendConfig gpu_config =
+            std::move(user->backend_config<GpuBackendConfig>()).value();
+        const GemmBackendConfig& config = gpu_config.gemm_backend_config();
```
Same nit as above: maybe remove extraneous lines and just use the arrow syntax?

akuegel commented:
//xla/service:xla_aot_compile_gpu_test is not adjusted yet. I tried to adjust it myself, but couldn't get it working so far. I would appreciate if you can try as well :)

Tixxx commented:
> //xla/service:xla_aot_compile_gpu_test is not adjusted yet. I tried to adjust it myself, but couldn't get it working so far. I would appreciate if you can try as well :) So one thing I tried was changing xla_aot_compile_test_gpu_target_config.prototxt but that alone was not enough.

Sure, let me try adjusting that. I'd need to do a rebase and resolve the conflicts

cezheng commented:
Some folks are actively changing this hlo_instruction.h file recently and caused some conflicts https://github.com/openxla/xla/commit/d664243f5b95c6119b7c250b040538327a28585e

Let's merge this sooner so we don't run into more of these.

Tixxx commented:
> > //xla/service:xla_aot_compile_gpu_test is not adjusted yet. I tried to adjust it myself, but couldn't get it working so far. I would appreciate if you can try as well :) So one thing I tried was changing xla_aot_compile_test_gpu_target_config.prototxt but that alone was not enough.
> 
> Sure, let me try adjusting that. I'd need to do a rebase and resolve the conflicts

I have been getting this error when trying to run that test:
```
xla/service/xla_aot_compile_gpu_test.cc:240:28: error: call of overloaded ‘Run(<brace-enclosed initializer list>, xla::ExecutableRunOptions&)’ is ambiguous
  240 |       local_executable->Run({&array1, &array2}, executable_run_options));

```
This even happens on openxla main branch so i dont think it's specific to my change.
I added changes to xla_aot_compile_test_gpu_target_config.prototxt in this pr but still am not able run it yet.
